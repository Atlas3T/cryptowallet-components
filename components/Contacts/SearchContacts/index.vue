<template>
  <section class="search-contacts">
    <div class="flex-wrapper">
      <div class="search-input-wrapper">
        <q-input
          v-model="searchQueryString"
          inverted
          clearable
          float-label="Search contacts"
          color="blueish"
          @focus="isSearching(true)"
          @blur="isSearching(false)"
        />

        <div
          v-show="isSearchingContacts"
          class="cancel-search-wrapper"
        >
          <q-btn
            color="secondary"
            label="Cancel"
            @click="cancelSearch"
          />
        </div>
      </div>
    </div>
  </section>
</template>

<script>
import { mapState } from 'vuex';

export default {
  name: 'SearchContacts',
  data() {
    return {
    };
  },

  computed: {
    ...mapState({
      isSearchingContacts: (state) => {
        return state.search.isSearchingContacts;
      },
      searchingContactsQueryString: (state) => {
        return state.search.searchingContactsQueryString;
      },
    }),

    searchQueryString: {

      /**
       * Use a computed value getter to keep the input value in sync with the store
       * @return {String}
       */
      get() {
        return this.$store.state.search.searchingContactsQueryString;
      },

      /**
       * Use a computed value setter to update the store if input value changes
       * @param {String} value The new input value
       */
      set(value) {
        this.$store.dispatch('search/updateSearchContactsQueryString', { value });
      },
    },

  },

  methods: {

    /**
     * Updates the store search/isSearchingContacts state
     * @param  {Boolean} value Either is searching or isn't
     * @return {Boolean}
     */
    isSearching(value) {
      this.$store.dispatch('search/updateIsSearchingContacts', { value });
    },

    /**
     * Will be called if user clicks the CANCEL button while searching contacts
     * Will dispatch actions needed to reset the input value and reset the UI state
     */
    cancelSearch() {
      this.$store.dispatch('search/updateIsSearchingContacts', { value: false });
      this.$store.dispatch('search/updateSearchContactsQueryString', { value: '' });
    },
  },
};
</script>

<style>
.search-contacts {
  transition: height 250ms ease-in-out;
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 5em;
}

.search-input-wrapper {
  width: 90vw;
  display: flex;
}

.search-input-wrapper > .q-input {
  width: 60%;
  flex-grow: 1;
}

.cancel-search-wrapper > .q-btn {
  margin-left: 1rem;
  height: 100%;
}

.search-input-wrapper .column {
  padding: 0;
}
</style>
