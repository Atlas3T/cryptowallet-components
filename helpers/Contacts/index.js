import Vue from 'vue';

/**
 * Handles the contacts import and saves them in the store
 */
export default function Contacts() {
  this.vm = new Vue();
  this.import = (contacts) => {
    this.contacts = contacts;
    this.vm.$once('contactsImported', () => { return this.success(); });
    this.vm.$once('contactsImportFailure', () => { return this.failure(); });
    this.vm.$once('contactsImportError', (event) => { return this.error(event.payload); });

    try {
      const options = new ContactFindOptions();
      options.multiple = true;
      const fields = ['*'];
      navigator.contacts.find(fields, (contactsArray) => {
        this.contacts.contacts = contactsArray;
        this.vm.$emit('contactsImported');
      }, () => {
        this.vm.$emit('contactsImportFailure');
      }, options);
    } catch (err) {
      this.vm.$emit('contactsImportError', { payload: err });
    }

    return this;
  };

  this.on = (event, callback) => {
    this[event] = callback;
    return this;
  };
}
