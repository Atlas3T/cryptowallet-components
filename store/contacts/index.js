export default {
  namespaced: true,
  state: {
    contacts: [
      {
        id: 1,
        displayName: 'Terry Tibbs',
        address: 'ncousohfodicwjdclwdnc',
      },
    ],
  },
};
