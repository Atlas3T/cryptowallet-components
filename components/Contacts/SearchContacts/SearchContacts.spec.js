import { shallowMount, mount } from '@vue/test-utils';
import SearchContacts from '@/components/Contacts/SearchContacts';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';

describe('SearchContacts.vue', () => {
  let storeMocks;
  let wrapper;
  let store;

  function wrapperInit (options) {
    return shallowMount(SearchContacts, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    wrapper = wrapperInit({ i18n, localVue, store: storeMocks.store });
    store = wrapper.vm.$store;
  }

  beforeEach(() => storeInit());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a section with .search-contacts class', () => {
    expect(wrapper.contains('section.search-contacts')).toBe(true);
  });

  it('renders an input', () => {
    expect(wrapper.contains('qinput-stub')).toBe(true);
  });

  it('displays input value using stores state search.searchingContactsQueryString', () => {
    wrapper = mount(SearchContacts, { i18n, localVue, store: storeMocks.store, sync: false });
    store.state.search.searchingContactsQueryString = 'hello there';
    localVue.nextTick(() => {
      expect(wrapper.find('input').element.value).toBe('hello there');
    });
  });

  it('dispatches search/updateSearchContactsQueryString with correct value on input change', () => {
    wrapper = mount(SearchContacts, { i18n, localVue, store: storeMocks.store });
    wrapper.find('input').setValue('test value');
    expect(storeMocks.actions.updateSearchContactsQueryString).toHaveBeenCalled();
    expect(storeMocks.actions.updateSearchContactsQueryString.mock.calls[0][1].value === 'test-value');
  });

  it('dispatches search/updateIsSearchingContacts with value true on input focus', () => {
    wrapper = mount(SearchContacts, { i18n, localVue, store: storeMocks.store });
    wrapper.find('input').trigger('focus');
    expect(storeMocks.actions.updateIsSearchingContacts).toHaveBeenCalled();
    expect(storeMocks.actions.updateIsSearchingContacts.mock.calls[0][1].value === true);
  });

  it('dispatches search/updateIsSearchingContacts with value false on input blur', () => {
    wrapper = mount(SearchContacts, { i18n, localVue, store: storeMocks.store });
    wrapper.find('input').trigger('blur');
    expect(storeMocks.actions.updateIsSearchingContacts).toHaveBeenCalled();
    expect(storeMocks.actions.updateIsSearchingContacts.mock.calls[0][1].value === false);
  });

  it('shows the cancel button if store has search/isSearchingContacts === true', () => {
    expect(store.state.search.isSearchingContacts).toBe(false);
    expect(wrapper.find('.cancel-search-wrapper').isVisible()).toBe(false);
    store.state.search.isSearchingContacts = true;
    expect(wrapper.find('.cancel-search-wrapper').isVisible()).toBe(true);

  });

  it('dispatches correct actions on the cancel button click', () => {
    wrapper = mount(SearchContacts, { i18n, localVue, store: storeMocks.store, sync: false });
    store.state.search.isSearchingContacts = true;
    wrapper.find('.cancel-search-wrapper button').trigger('click');
    expect(storeMocks.actions.updateIsSearchingContacts).toHaveBeenCalled();
    expect(storeMocks.actions.updateIsSearchingContacts.mock.calls[0][1].value === false);
    expect(storeMocks.actions.updateSearchContactsQueryString).toHaveBeenCalled();
    expect(storeMocks.actions.updateSearchContactsQueryString.mock.calls[0][1].value === '');
  });
});
