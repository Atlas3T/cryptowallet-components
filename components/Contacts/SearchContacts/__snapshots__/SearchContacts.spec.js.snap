// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContacts.vue renders and matches snapshot 1`] = `
<section
  class="search-contacts"
>
  <div
    class="flex-wrapper"
  >
    <div
      class="search-input-wrapper"
    >
      <qinput-stub
        clearable=""
        color="blueish"
        float-label="Search contacts"
        inverted=""
        type="text"
        value=""
      />
       
      <div
        class="cancel-search-wrapper"
        style="display: none;"
      >
        <qbtn-stub
          color="secondary"
          label="Cancel"
        />
      </div>
    </div>
  </div>
</section>
`;
